### PR TITLE
Added size support to Network tab

### LINKED ISSUE #6165: Show size of network request in Network tab
It would be helpful if the Network tab showed the size of the request and response in bytes. Right now, it only displays timing information:

Screenshot 2023-08-07 at 10 35 40 AM

### PR DESCRIPTION
**Pull request by Victowolf**
## Overview

This PR adds support for displaying response payload size in the Network tab
and fixes #6165.

It introduces a new **"Size"** column in the network requests table and displays response size in the **Overview panel** of the request inspector.

---

### 1. Data Model Updates

**File:**
`packages/devtools_app/lib/src/screens/network/network_model.dart`

* Added two new getters to the `NetworkRequest` base class:

  * `requestBytes`
  * `responseBytes`
* Implemented these getters in the `Socket` class using:

  * `writeBytes` : request size
  * `readBytes` : response size

**Purpose:**
Expose byte-level data in a unified way for all network request types.

---

### 2. HTTP Data Handling

**File:**
`packages/devtools_app/lib/src/shared/http/http_request_data.dart`

* Added logic to extract response size using the `content-length` header
* Handles both `String` and `List<String>` header formats

**Purpose:**
Provide response size for HTTP requests when available, without requiring changes to the Dart VM.

---

### 3. Shared Utility

**File:**
`packages/devtools_app/lib/src/screens/network/utils/http_utils.dart`

* Moved `formatBytes` into a reusable utility function
* Uses **decimal (base-10) units** (`kB`, `MB`) to align with Chrome DevTools
* Handles null and negative values safely

**Purpose:**
Ensure consistent formatting across the network table and inspector views.

---

### 4. Network Table UI

**File:**
`packages/devtools_app/lib/src/screens/network/network_screen.dart`

* Added a new column: **"Size"**
* Displays formatted response size
* Shows `-` when size is unavailable

---

### 5. Request Inspector (Overview Panel)

**File:**
`packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart`

* Added a new row:

  * **Response Size**
* Uses shared `formatBytes` utility

---

### 6. Tests

* Added unit tests for:

  * `formatBytes` utility in `http_utils_test.dart`
  * `responseBytes` parsing logic in `http_request_data.dart`
* Covers edge cases including:

  * string and list headers
  * invalid values
  * null handling

---

## Why request size is not included

Request size is not reliably available for HTTP requests due to limitations in the current DevTools and VM service APIs:

* The Dart VM does not expose request payload size in `HttpProfileRequest`
* HTTP request bodies are not always accessible or fully captured
* Headers such as `content-length` are often absent for outgoing requests
* Streaming and chunked requests complicate accurate measurement

While socket-level request size (`writeBytes`) is available, it is not consistently applicable to HTTP requests.

Therefore, including request size would require changes in the Dart SDK / VM layer.

This PR focuses on **response size**, which can be reliably determined using:

* Socket `readBytes`
* HTTP `content-length` header (when present)

---

## Screenshot

Screenshot 2026-03-27 233804

---

## Future Work

* Add request size support when VM-level data becomes available
* Introduce separate request/response size columns
* Improve accuracy via VM instrumentation

---

### General checklist

* [x] I read the Contributor Guide
* [x] I read the Tree Hygiene guidelines
* [x] I followed the Flutter Style Guide
* [x] I signed the CLA
* [x] I updated relevant documentation

---

### Issues checklist

* [x] This PR fixes #6165

---

### Tests checklist

* [x] Added unit tests for new functionality

---

### AI-tooling checklist

* [x] I used AI tooling responsibly and verified all generated content

---

### Feature-change checklist

* [x] This PR changes DevTools UI
* [x] Added entry to `NEXT_RELEASE_NOTES.md`
* [x] Included screenshots
* [x] Verified changes locally

### CODE REVIEW

File: packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart [hunk 1]
```diff
@@ -669,6 +677,15 @@ class NetworkRequestOverviewView extends StatelessWidget {
     ];
   }
 
+  String _formatBytes(int? bytes) {
```
Comment (srawlins) [inline packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart]: Can you add a comment explaining what the output might look like? I'm thinking specifically about any rounding.

File: packages/devtools_app/lib/src/screens/network/network_model.dart [hunk 2]
```diff
@@ -29,6 +29,9 @@ abstract class NetworkRequest
 
   int? get port;
 
+  int? get requestBytes => null;
```
Comment (srawlins) [inline packages/devtools_app/lib/src/screens/network/network_model.dart:32]: Can these be left abstract?

**Review by srawlins — APPROVED** [review]
Looks great, thank you!

File: packages/devtools_app/lib/src/screens/network/network_model.dart (hunk 2 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/screens/network/network_model.dart:32]: I kept them as non-abstract getters with a default `null` implementation because not all `NetworkRequest` types can reliably provide byte information. 
Making them abstract would require every subclass to implement these getters, even when the data is not available.
and also using a default `null` allows subclasses like `Socket` and `DartIOHttpRequestData` to override when data is available, while others can safely fallback.

**Comment by srawlins** [comment]
Merge in the master branch to resolve that merge conflict.

File: packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart (hunk 1 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart]: Done. 👍

File: packages/devtools_app/lib/src/screens/network/network_model.dart (hunk 2 above)
Comment (srawlins) [reply packages/devtools_app/lib/src/screens/network/network_model.dart:32]: OK that's fair.

**Review by srawlins — APPROVED** [review]

**Comment by Victowolf** [comment]
Hi @srawlins  
I’ve pushed fixes based on the failed tests. 
Could you please approve the workflow so CI can run again?
Thanks!

**Review by srawlins — APPROVED** [review]

**Comment by Victowolf** [comment]
Hi @srawlins,

One test failed: `ubuntu-latest test_ddc`

The failing `test_ddc `appears to be flaky (`asyncEval` returning null intermittently).
My changes were limited to the Network UI, but I fixed a parameter mismatch in `ResponseSizeColumn`, which i think was the cause of the failed test.
Could you please re-approve so CI can run again? 

Thanks!

File: packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart [hunk 3]
```diff
@@ -669,6 +678,20 @@ class NetworkRequestOverviewView extends StatelessWidget {
     ];
   }
 
+  // Output Formats:
+  // - 512 → "512 B"
+  // - 2048 → "2.0 KB"
+  // - 1048576 → "1.0 MB"
+  // Values are rounded to one decimal place for KB and MB.
+  String _formatBytes(int? bytes) {
```
Comment (elliette) [inline packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart]: I think we should should use Chrome DevTools network panel as a model for our network panel as much as possible. It looks like they decided to standardize on using `kB` (1000 bytes) instead of KiB (1024 bytes) as their unit of measurement **[1]**. We should do the same unless there is a good reason not to.

[1] https://developer.chrome.com/blog/new-in-devtools-88#consistent-kb

File: packages/devtools_app/lib/src/screens/network/network_screen.dart [hunk 4]
```diff
@@ -371,6 +371,7 @@ class NetworkRequestsTable extends StatelessWidget {
     statusColumn,
     typeColumn,
     durationColumn,
+    const ResponseSizeColumn(),
```
Comment (elliette) [inline packages/devtools_app/lib/src/screens/network/network_screen.dart]: [nit] add this to the list of static const fields above

File: packages/devtools_app/lib/src/shared/http/http_request_data.dart [hunk 5]
```diff
@@ -227,6 +227,27 @@ class DartIOHttpRequestData extends NetworkRequest {
     return connectionInfo != null ? connectionInfo[_localPortKey] : null;
   }
 
+  @override
+  int? get responseBytes {
+    final headers = responseHeaders;
```
Comment (elliette) [inline packages/devtools_app/lib/src/shared/http/http_request_data.dart]: I don't think we need to assign `responseHeaders` to a new variable here, can simply write:

```
 final contentLength = responseHeaders?['content-length'];
```

File: packages/devtools_app/lib/src/screens/network/network_screen.dart [hunk 6]
```diff
@@ -415,8 +444,8 @@ class AddressColumn extends ColumnData<NetworkRequest>
       );
 
   @override
-  String getValue(NetworkRequest dataObject) {
-    return dataObject.uri;
+  String getValue(NetworkRequest data) {
```
Comment (elliette) [inline packages/devtools_app/lib/src/screens/network/network_screen.dart]: I don't think this change is necessary

**Review by elliette — CHANGES_REQUESTED** [review]

File: packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart (hunk 3 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart]: yes.. right
i didn't knew about this, thanks for pointing it out.

File: packages/devtools_app/lib/src/screens/network/network_screen.dart (hunk 4 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/screens/network/network_screen.dart]: will update it...

File: packages/devtools_app/lib/src/shared/http/http_request_data.dart (hunk 5 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/shared/http/http_request_data.dart]: yes , we can merge it in a single line, 
Thanks for this

File: packages/devtools_app/lib/src/screens/network/network_screen.dart (hunk 6 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/screens/network/network_screen.dart]: yes my bad, i changed it trying to fix a test fail, i will revert it back.

File: packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart (hunk 3 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/screens/network/network_request_inspector_views.dart]: updated 👍

File: packages/devtools_app/lib/src/screens/network/network_screen.dart (hunk 4 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/screens/network/network_screen.dart]: updated 👍

File: packages/devtools_app/lib/src/shared/http/http_request_data.dart (hunk 5 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/shared/http/http_request_data.dart]: updated 👍

File: packages/devtools_app/lib/src/screens/network/network_screen.dart (hunk 6 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/screens/network/network_screen.dart]: reverted 👍

File: packages/devtools_app/lib/src/screens/network/network_screen.dart [hunk 7]
```diff
@@ -405,6 +407,35 @@ class NetworkRequestsTable extends StatelessWidget {
   }
 }
 
+// Output Formats:
+// - 512 → "512 B"
+// - 2000 → "2.0 kB"
+// - 1000000 → "1.0 MB"
+// Values are rounded to one decimal place for kB and MB.
+// Uses decimal (base-10) units to match Chrome DevTools.
+String _formatBytes(int? bytes) {
```
Comment (elliette) [inline packages/devtools_app/lib/src/screens/network/network_screen.dart]: Move `formatBytes` into [`http_utils.dart` ](https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/screens/network/utils/http_utils.dart) so it can be used here and in `network_request_inspector_view.dart`, and add a test

File: packages/devtools_app/release_notes/NEXT_RELEASE_NOTES.md [hunk 8]
```diff
@@ -39,6 +39,9 @@ TODO: Remove this section if there are not any updates.
 
 ## Network profiler updates
 
+- Added response size ("Res Size") column to the Network tab and displayed response size in the request inspector overview. -
```
Comment (elliette) [inline packages/devtools_app/release_notes/NEXT_RELEASE_NOTES.md]: Can remove ("Res Size"), i.e. "Added response size column to..."

File: packages/devtools_app/lib/src/shared/http/http_request_data.dart [hunk 9]
```diff
@@ -227,6 +227,26 @@ class DartIOHttpRequestData extends NetworkRequest {
     return connectionInfo != null ? connectionInfo[_localPortKey] : null;
   }
 
+  @override
+  int? get responseBytes {
```
Comment (elliette) [inline packages/devtools_app/lib/src/shared/http/http_request_data.dart:255]: Please add a test for this

**Review by elliette — CHANGES_REQUESTED** [review]
Thanks! A few changes but looks good!

File: packages/devtools_app/lib/src/screens/network/network_screen.dart (hunk 7 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/screens/network/network_screen.dart]: Done 👍

File: packages/devtools_app/release_notes/NEXT_RELEASE_NOTES.md (hunk 8 above)
Comment (Victowolf) [reply packages/devtools_app/release_notes/NEXT_RELEASE_NOTES.md]: on it.....

Comment (Victowolf) [reply packages/devtools_app/release_notes/NEXT_RELEASE_NOTES.md]: Updated it 👍

File: packages/devtools_app/lib/src/shared/http/http_request_data.dart (hunk 9 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/shared/http/http_request_data.dart:255]: Sure, added tests in `/test/shared/http/http_request_data_test.dart`

File: packages/devtools_app/lib/src/screens/network/network_screen.dart (hunk 7 above)
Comment (Victowolf) [reply packages/devtools_app/lib/src/screens/network/network_screen.dart]: Added the tests in `test/shared/http/http_utils_test.dart`.

**Review by elliette — APPROVED** [review]
Thank you, looks good!

**Comment by Victowolf** [comment]
Hi, @elliette @srawlins 
Could anyone also please approve the workflow so CI can run?

**Review by elliette — CHANGES_REQUESTED** [review]
Please fix the failing tests in `http_request_data_test.dart`, thank you!

**Comment by Victowolf** [comment]
I've pushed the fixes for the failed tests. Could you please approve the workflow so the CI can run again? 
Thanks!

**Comment by Victowolf** [comment]
I found it
The issue was that the `response` map in `baseJson` was missing the required `redirects` field, which caused a null-bang crash inside `HttpProfileResponseData._fromJson`. Added `'redirects': <Map<String, dynamic>>[]` to resolve it.

**Comment by Victowolf** [comment]
The remaining failures (`eval_integration_test` timeout and `integration_dart2js` DWDS `WipError`) seem unrelated to my changes and also appear to be CI Flakes. As this is my first contribution, I'm not sure how to handle these.....
any advise?

Thanks!

**Comment by elliette** [comment]
> The remaining failures (`eval_integration_test` timeout and `integration_dart2js` DWDS `WipError`) seem unrelated to my changes and also appear to be CI Flakes. As this is my first contribution, I'm not sure how to handle these..... any advise?
> 
> Thanks!

Let me try re-running them to deflake

**Comment by Victowolf** [comment]
the failed error is also a flake.
re-running might solve it too.....

**Comment by Victowolf** [comment]
All Tests passed.....
need final approval for merging.

Thanks for the guidance.....

**Comment by Victowolf** [comment]
Hi @elliette 
All requested changes have been addressed and CI is now passing.
Could you please take another look and re-approve if everything looks good? 
Thanks!

**Comment by elliette** [comment]
You will need to resync now that `NEXT_RELEASE_NOTES` have changed. Then hopefully can get this merged. Thanks!

**Comment by Victowolf** [comment]
Hi @elliette
I’ve synced with latest master and resolved conflicts. 
It needs approving for ci workflows to run..
Thanks..!!

File: pubspec.lock [hunk 10]
```diff
@@ -546,10 +546,10 @@ packages:
     dependency: transitive
     description:
       name: meta
-      sha256: "1741988757a65eb6b36abe716829688cf01910bbf91c34354ff7ec1c3de2b349"
+      sha256: "23f08335362185a5ea2ad3a4e597f1375e78bce8a040df5c600c8d3552ef2394"
```
Comment (srawlins) [inline pubspec.lock]: I think revert this change.

**Review by srawlins — APPROVED** [review]

File: pubspec.lock (hunk 10 above)
Comment (Victowolf) [reply pubspec.lock]: Reverted 👍
This seems to have been staged unintentionally during the rebase. I’ll be more careful with staged files going forward.

**Comment by Victowolf** [comment]
Hi @srawlins,
I have reverted unnecessary changes....
Now it needs approving for ci workflows to run..
Thanks..!!

**Comment by Victowolf** [comment]
Hi @srawlins,
The failed tests are flaky...
Rerunning them might solve it like earlier.
Thanks..

**Comment by elliette** [comment]
Please sync with master branch to resolve conflicts, thanks!

**Comment by Victowolf** [comment]
@srawlins @elliette 
synced & resolved conflicts....
It needs approving for ci workflows to run..
Thanks..!!

**Review by srawlins — APPROVED** [review]

**Comment by Victowolf** [comment]
Looks like a flaky DevTools integration test (DWDS / “Promise was collected”), not related to present code changes.
Could we re-run the the test once to confirm?

**Comment by Victowolf** [comment]
@srawlins 
I think I may have found the cause, the `responseBytes` getter was indirectly accessing headers in a way that could interfere with the request lifecycle timing. I’ve updated it to be more direct/non-intrusive.

Need approval for CI.

**Review by srawlins — APPROVED** [review]

**Review by elliette — APPROVED** [review]
